{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 1000 Genomes PCA "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Chris Porras\n",
    "A notebook to parse 1000 genomes data, convert SNPs to PCA-ble format, and vizualize."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits import mplot3d\n",
    "import plotly.express as px\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/chris/anaconda3/lib/python3.6/site-packages/h5py/__init__.py:36: FutureWarning:\n",
      "\n",
      "Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# import scikit-allel\n",
    "import allel"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Test if increase memory alloc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Couldn't find Spark, make sure SPARK_HOME env is set or Spark is in an expected location (e.g. from homebrew installation).",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-3-cd5383b0b018>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mos\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mfindspark\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mfindspark\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0minit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      4\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mpyspark\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/findspark.py\u001b[0m in \u001b[0;36minit\u001b[0;34m(spark_home, python_path, edit_rc, edit_profile)\u001b[0m\n\u001b[1;32m    120\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    121\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mspark_home\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 122\u001b[0;31m         \u001b[0mspark_home\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfind\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    123\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    124\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mpython_path\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/findspark.py\u001b[0m in \u001b[0;36mfind\u001b[0;34m()\u001b[0m\n\u001b[1;32m     31\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     32\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mspark_home\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 33\u001b[0;31m         raise ValueError(\"Couldn't find Spark, make sure SPARK_HOME env is set\"\n\u001b[0m\u001b[1;32m     34\u001b[0m                          \" or Spark is in an expected location (e.g. from homebrew installation).\")\n\u001b[1;32m     35\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Couldn't find Spark, make sure SPARK_HOME env is set or Spark is in an expected location (e.g. from homebrew installation)."
     ]
    }
   ],
   "source": [
    "import os\n",
    "import findspark\n",
    "findspark.init()\n",
    "\n",
    "import pyspark\n",
    "import random\n",
    "\n",
    "memory = '4g'\n",
    "pyspark_submit_args = ' --driver-memory ' + memory + ' pyspark-shell'\n",
    "os.environ[\"PYSPARK_SUBMIT_ARGS\"] = pyspark_submit_args\n",
    "\n",
    "sc = pyspark.SparkContext(appName=\"Pi\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Reading 1000 genomes chr21 genotypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "GT = allel.read_vcf('../../data/vcf/ALL.chr21.phase3_shapeit2_mvncall_'\n",
    "                     +'integrated_v5a.20130502.genotypes.vcf.gz',\n",
    "                    fields = ['calldata/GT'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "GT = GT['calldata/GT']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "ALT = np.load('alternate_seq.npy')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "REF = np.load('reference_seq.npy')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample_ids = np.load('sample_ids.npy')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Construct genotype array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# gt_arr = allel.GenotypeArray(GT)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([5.30532465e+09, 2.28747666e+08, 0.00000000e+00, 2.31172700e+06,\n",
       "        0.00000000e+00, 1.29536000e+05, 1.56790000e+04, 0.00000000e+00,\n",
       "        4.93300000e+03, 1.10000000e+02]),\n",
       " array([0. , 0.6, 1.2, 1.8, 2.4, 3. , 3.6, 4.2, 4.8, 5.4, 6. ]),\n",
       " <a list of 10 Patch objects>)"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAW4AAAEDCAYAAAAVyO4LAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvhp/UCwAAC+tJREFUeJzt3H2IZXUdx/HPp92tzAeE9ibSqlMRhkipDUYtiJnFlqKVBEkKhbD/VCglov0TBVFBiH/0AItuKvmA+QChZS7qsgr5MONTbqsRttG2xoyI6fZHoX76Y87ats7sPTsz5577nd4vGObOnbP3fA/Lvjn7u+dcJxEAoI639D0AAODgEG4AKIZwA0AxhBsAiiHcAFAM4QaAYjoLt+3NtmdsP91i2+Ns32v7Kdtbba/rai4AqK7LM+5rJW1oue2PJF2f5IOSvivp+10NBQDVdRbuJNskvbjvc7bfZ/tu29O2H7D9geZXJ0i6t3l8v6Rzu5oLAKob9Rr3JklfT/JhSZdK+mnz/JOSzmsef07S4bbfOeLZAKCE1aPake3DJH1M0i9t7336bc33SyX92PaXJW2T9DdJr45qNgCoZGTh1tzZ/UtJTtr/F0l2S/q89Ebgz0vyjxHOBgBljGypJMnLkv5s+wuS5Dkfah6vtb13liskbR7VXABQTZeXA94k6XeSjre9y/ZFkr4k6SLbT0rarv++CXm6pGdt/1HSUZK+19VcAFCd+VhXAKiFOycBoJhO3pxcu3ZtJiYmunhpAFiRpqenX0gyaLNtJ+GemJjQ1NRUFy8NACuS7b+03ZalEgAohnADQDGEGwCKIdwAUAzhBoBiCDcAFEO4AaAYwg0AxRBuAChmlJ/H3crE5Xf1st+dPzirl/0CwMHijBsAiiHcAFAM4QaAYgg3ABRDuAGgGMINAMUQbgAoptV13LZ3SnpF0muSXk0y2eVQAICFHcwNOB9P8kJnkwAAWmGpBACKaRvuSLrH9rTtjV0OBAA4sLZLJeuT7Lb9LklbbD+TZNu+GzRB3yhJxx577DKPCQDYq9UZd5LdzfcZSXdIOnWebTYlmUwyORgMlndKAMAbhobb9qG2D9/7WNKnJD3d9WAAgPm1WSo5StIdtvduf2OSuzudCgCwoKHhTvKcpA+NYBYAQAtcDggAxRBuACiGcANAMYQbAIoh3ABQDOEGgGIINwAUQ7gBoBjCDQDFEG4AKIZwA0AxhBsAiiHcAFAM4QaAYgg3ABRDuAGgGMINAMUQbgAohnADQDGEGwCKIdwAUAzhBoBiCDcAFEO4AaAYwg0AxRBuACiGcANAMYQbAIoh3ABQDOEGgGIINwAU0zrctlfZftz2nV0OBAA4sIM5475Y0o6uBgEAtNMq3LbXSTpL0tXdjgMAGKbtGfdVki6T9PpCG9jeaHvK9tTs7OyyDAcAeLOh4bZ9tqSZJNMH2i7JpiSTSSYHg8GyDQgA+F9tzrjXSzrH9k5JN0s6w/YvOp0KALCgoeFOckWSdUkmJH1R0n1JLuh8MgDAvLiOGwCKWX0wGyfZKmlrJ5MAAFrhjBsAiiHcAFAM4QaAYgg3ABRDuAGgGMINAMUQbgAohnADQDGEGwCKIdwAUAzhBoBiCDcAFEO4AaAYwg0AxRBuACiGcANAMYQbAIoh3ABQDOEGgGIINwAUQ7gBoBjCDQDFEG4AKIZwA0AxhBsAiiHcAFAM4QaAYgg3ABRDuAGgGMINAMUMDbftt9t+xPaTtrfb/s4oBgMAzG91i23+JemMJHtsr5H0oO3fJHmo49kAAPMYGu4kkbSn+XFN85UuhwIALKzVGrftVbafkDQjaUuSh+fZZqPtKdtTs7Ozyz0nAKDRKtxJXktykqR1kk61feI822xKMplkcjAYLPecAIDGQV1VkuQlSVslbehkGgDAUG2uKhnYPrJ5fIikMyU90/VgAID5tbmq5GhJ19lepbnQ35Lkzm7HAgAspM1VJU9JOnkEswAAWuDOSQAohnADQDGEGwCKIdwAUAzhBoBiCDcAFEO4AaAYwg0AxRBuACiGcANAMYQbAIoh3ABQDOEGgGIINwAUQ7gBoBjCDQDFEG4AKIZwA0AxhBsAiiHcAFAM4QaAYgg3ABRDuAGgGMINAMUQbgAohnADQDGEGwCKIdwAUAzhBoBiCDcAFDM03LaPsX2/7R22t9u+eBSDAQDmt7rFNq9K+maSx2wfLmna9pYkf+h4NgDAPIaecSd5PsljzeNXJO2Q9O6uBwMAzO+g1rhtT0g6WdLD8/xuo+0p21Ozs7PLMx0A4E1ah9v2YZJuk3RJkpf3/32STUkmk0wOBoPlnBEAsI9W4ba9RnPRviHJ7d2OBAA4kDZXlVjSNZJ2JLmy+5EAAAfS5ox7vaQLJZ1h+4nm6zMdzwUAWMDQywGTPCjJI5gFANACd04CQDGEGwCKIdwAUAzhBoBiCDcAFEO4AaAYwg0AxRBuACiGcANAMYQbAIoh3ABQDOEGgGIINwAUQ7gBoBjCDQDFEG4AKIZwA0AxhBsAiiHcAFAM4QaAYgg3ABRDuAGgGMINAMUQbgAohnADQDGEGwCKIdwAUAzhBoBiCDcAFEO4AaCYoeG2vdn2jO2nRzEQAODA2pxxXytpQ8dzAABaGhruJNskvTiCWQAALSzbGrftjbanbE/Nzs4u18sCAPazbOFOsinJZJLJwWCwXC8LANgPV5UAQDGEGwCKaXM54E2SfifpeNu7bF/U/VgAgIWsHrZBkvNHMQgAoB2WSgCgGMINAMUQbgAohnADQDGEGwCKIdwAUAzhBoBiCDcAFEO4AaAYwg0AxRBuACiGcANAMYQbAIoh3ABQDOEGgGIINwAUQ7gBoBjCDQDFEG4AKIZwA0AxhBsAiiHcAFAM4QaAYgg3ABSzuu8BxsXE5Xf1tu+dPzirt30DqIczbgAohnADQDGEGwCKIdwAUAzhBoBiWoXb9gbbz9r+k+3Lux4KALCwoeG2vUrSTyR9WtIJks63fULXgwEA5tfmjPtUSX9K8lySf0u6WdK53Y4FAFhImxtw3i3pr/v8vEvSR/bfyPZGSRubH/fYfnaRM62V9MIi/+y4aXUs/uEIJlma/7u/kyJWyrGslOOQlnYsx7XdsE24Pc9zedMTySZJm9rueMGd2VNJJpf6OuNgpRzLSjkOiWMZRyvlOKTRHUubpZJdko7Z5+d1knZ3Mw4AYJg24X5U0vttv8f2WyV9UdKvuh0LALCQoUslSV61/TVJv5W0StLmJNs7nGnJyy1jZKUcy0o5DoljGUcr5TikER2LkzctVwMAxhh3TgJAMYQbAIoZm3CvlNvqbW+2PWP76b5nWSrbx9i+3/YO29ttX9z3TItl++22H7H9ZHMs3+l7pqWwvcr247bv7HuWpbC90/bvbT9he6rveZbC9pG2b7X9TPNv5qOd7Wsc1rib2+r/KOmTmrv88FFJ5yf5Q6+DLYLt0yTtkXR9khP7nmcpbB8t6egkj9k+XNK0pM8W/XuxpEOT7LG9RtKDki5O8lDPoy2K7W9ImpR0RJKz+55nsWzvlDSZpPwNOLavk/RAkqubK/DekeSlLvY1LmfcK+a2+iTbJL3Y9xzLIcnzSR5rHr8iaYfm7qQtJ3P2ND+uab76P2tZBNvrJJ0l6eq+Z8Ec20dIOk3SNZKU5N9dRVsan3DPd1t9yUCsVLYnJJ0s6eF+J1m8ZnnhCUkzkrYkqXosV0m6TNLrfQ+yDCLpHtvTzcdmVPVeSbOSft4sYV1t+9CudjYu4W51Wz36YfswSbdJuiTJy33Ps1hJXktykubu/j3VdrmlLNtnS5pJMt33LMtkfZJTNPfpo19tlhorWi3pFEk/S3KypH9K6uy9unEJN7fVj6lmPfg2STckub3veZZD81/YrZI29DzKYqyXdE6zNnyzpDNs/6LfkRYvye7m+4ykOzS3bFrRLkm79vlf3K2aC3knxiXc3FY/hpo39K6RtCPJlX3PsxS2B7aPbB4fIulMSc/0O9XBS3JFknVJJjT37+S+JBf0PNai2D60edNbzbLCpySVvBoryd8l/dX28c1Tn5DU2Zv4bT4dsHM93FbfGds3STpd0lrbuyR9O8k1/U61aOslXSjp983asCR9K8mve5xpsY6WdF1zBdNbJN2SpPSldCvAUZLumDs/0GpJNya5u9+RluTrkm5oTj6fk/SVrnY0FpcDAgDaG5elEgBAS4QbAIoh3ABQDOEGgGIINwAUQ7gBoBjCDQDF/Af7+6v9Th/PzwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.hist(np.ravel(GT))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Convert binomial SNP coding to monomial with number of variants ( 0 = ref, 1 = var)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "GT[GT> 0] = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "ename": "MemoryError",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mMemoryError\u001b[0m                               Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-14-91a6c3bc4c82>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msave\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'summarized_genotypes'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mGT\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/numpy/core/fromnumeric.py\u001b[0m in \u001b[0;36msum\u001b[0;34m(a, axis, dtype, out, keepdims)\u001b[0m\n\u001b[1;32m   1880\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0msum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1881\u001b[0m     return _methods._sum(a, axis=axis, dtype=dtype,\n\u001b[0;32m-> 1882\u001b[0;31m                          out=out, **kwargs)\n\u001b[0m\u001b[1;32m   1883\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1884\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/lib/python3.6/site-packages/numpy/core/_methods.py\u001b[0m in \u001b[0;36m_sum\u001b[0;34m(a, axis, dtype, out, keepdims)\u001b[0m\n\u001b[1;32m     30\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     31\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0m_sum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkeepdims\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 32\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mumr_sum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkeepdims\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     33\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     34\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0m_prod\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkeepdims\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mMemoryError\u001b[0m: "
     ]
    }
   ],
   "source": [
    "np.save('summarized_genotypes',np.sum(GT,axis=2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Perform PCA on entire sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.decomposition import PCA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "######### TAKES FOREVER!! ###########"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pca = PCA(n_components=4)\n",
    "PCs = pca.fit_transform(conv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "######################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Projections onto PCs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PC1 = conv.T@PCs[:,0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PC2 = conv.T@PCs[:,1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "PC3 = conv.T@PCs[:,2]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Get sample population location"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "key = pd.read_csv(\"20130606_g1k.ped\", sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "key.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Get indeces of rows for our data in the key"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "key_id = key.iloc[:,1]\n",
    "GT_id = callset[\"samples\"]\n",
    "idx = []\n",
    "for i in range(len(key_id)):\n",
    "    for j in range(len(GT_id)):\n",
    "        if key_id[i] == GT_id[j]:\n",
    "            idx += i"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pops = key.iloc[idx,6] # populations for our chr21 data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "super_pop = pd.read_csv(\"20131219.superpopulations.tsv\", sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "super_pop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pop_code = pd.read_csv(\"20131219.populations.tsv\", sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get indeces of 1000 genomes chr21 samples from key\n",
    "uni_pops = []\n",
    "for i in np.unique(pops):\n",
    "    uni_pops += [np.where(pop_code.iloc[:,1]==i)[0][0]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pop_arr = np.array(pops)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(len(np.unique(pops))):\n",
    "    pop_arr[np.where(pops == np.unique(pops)[i])[0]] = np.array(pop_code.iloc[uni_pops,2])[i]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Pull out indeces by superpopulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "AFR_idx= np.where(pop_arr == np.unique(pop_arr)[np.where(np.unique(pop_arr)==\"AFR\")[0][0]])[0]\n",
    "AMR_idx= np.where(pop_arr == np.unique(pop_arr)[np.where(np.unique(pop_arr)==\"AMR\")[0][0]])[0]\n",
    "EAS_idx= np.where(pop_arr == np.unique(pop_arr)[np.where(np.unique(pop_arr)==\"EAS\")[0][0]])[0]\n",
    "EUR_idx= np.where(pop_arr == np.unique(pop_arr)[np.where(np.unique(pop_arr)==\"EUR\")[0][0]])[0]\n",
    "SAS_idx= np.where(pop_arr == np.unique(pop_arr)[np.where(np.unique(pop_arr)==\"SAS\")[0][0]])[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Vizualize by superpopulation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pops_col = np.array(pops)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "col_pal = np.arange(len(np.unique(pop_arr)))\n",
    "for i in col_pal:\n",
    "    pops_col[np.where(pop_arr == np.unique(pop_arr)[i])[0]] = col_pal[i]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "from mpl_toolkits import mplot3d\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "%matplotlib inline \n",
    "\n",
    "fig = plt.figure(figsize=(15,15))\n",
    "\n",
    "ax = plt.axes(projection ='3d')\n",
    "\n",
    "\n",
    "samples = [AFR_idx,AMR_idx,EAS_idx,EUR_idx,SAS_idx]\n",
    "colors = np.arange(len(samples))\n",
    "\n",
    "for samp in samples:\n",
    "    ax.scatter3D(PC1[samp],PC2[samp],PC3[samp], label = pop_arr[samp][0])\n",
    "ax.set_xlabel('PC1', fontsize=30)\n",
    "ax.set_xticks([])\n",
    "ax.set_ylabel('PC2',fontsize=24)\n",
    "ax.set_yticks([])\n",
    "ax.set_zlabel(r'PC3', fontsize=24,rotation = 90)\n",
    "ax.set_zticks([])\n",
    "ax.set_title(\"1000 Genomes chr21 unfiltered\",fontsize=30,pad = 40)\n",
    "ax.legend(loc =4,prop={'size':20})\n",
    "fig.savefig(\"unfiltered_chr21_PCA.jpg\",quality=95)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ax.set_xlabel('PC1', fontsize=30)\n",
    "ax.set_xticks([])\n",
    "ax.set_ylabel('PC2',fontsize=24)\n",
    "ax.set_yticks([])\n",
    "ax.set_zlabel(r'PC3', fontsize=24,rotation = 90)\n",
    "ax.set_zticks([])\n",
    "ax.set_title(\"1000 Genomes chr21 unfiltered\",fontsize=30,pad = 40)\n",
    "ax.legend(loc =4,prop={'size':20})\n",
    "fig.savefig(\"unfiltered_chr21_PCA.jpg\",quality=95)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "plt.figure(figsize=(15,15))\n",
    "for samp in samples:\n",
    "    plt.scatter(PC1[samp],PC2[samp], label = pop_arr[samp][0])\n",
    "plt.xlabel(\"PC1\",fontsize=30,labelpad=20)\n",
    "plt.xticks([])\n",
    "plt.ylabel(\"PC2\",fontsize=30,labelpad=20)\n",
    "plt.yticks([])\n",
    "plt.legend(loc =4,prop={'size':20})\n",
    "plt.title(\"1000 Genomes chr21 unfiltered\",fontsize=30,pad = 40)\n",
    "\n",
    "plt.savefig(\"unfiltered_chr21_PC1PC2.jpg\",quality=95)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(15,15))\n",
    "for samp in samples:\n",
    "    plt.scatter(PC2[samp],PC3[samp], label = pop_arr[samp][0])\n",
    "plt.xlabel(\"PC2\",fontsize=30,labelpad=20)\n",
    "plt.xticks([])\n",
    "plt.ylabel(\"PC3\",fontsize=30,labelpad=20)\n",
    "plt.yticks([])\n",
    "plt.legend(loc =3,prop={'size':20})\n",
    "plt.title(\"1000 Genomes chr21 unfiltered\",fontsize=30,pad = 40)\n",
    "\n",
    "plt.savefig(\"unfiltered_chr21_PC2PC3.jpg\",quality=95)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
